README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# {cranology}: The CRAN Chronology

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
[![R-CMD-check](https://github.com/ThinkR-open/cranology/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/ThinkR-open/cranology/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

The goal of `{cranology}` is to provide tools to scrape data from CRAN
and MRAN websites as well as useful datasets to explore the evolution of
the number of packages on CRAN.

``` r
cranology::plot_cran_monthly_package_number()
```

<img src="man/figures/README-unnamed-chunk-2-1.png" width="100%" />

## Installation

You can install the development version of `{cranology}` with:

``` r
remotes::install_github("ThinkR-open/cranology")
```

## Examples

``` r
library(cranology)
```

### Datasets

All packages ever available on CRAN.

``` r
cran_packages_history
#> # A tibble: 37,823 × 10
#>    file_name    date                time  size  package_name last_archived      
#>    <chr>        <dttm>              <chr> <chr> <chr>        <dttm>             
#>  1 A3/          2015-08-16 21:05:00 21:05 -     A3           2015-08-16 21:05:00
#>  2 aaMI/        2010-07-30 12:17:00 12:17 -     aaMI         2010-07-30 12:17:00
#>  3 aaSEA/       2022-06-21 05:12:00 05:12 -     aaSEA        2022-06-21 05:12:00
#>  4 aba/         2022-03-27 06:29:00 06:29 -     aba          2022-03-27 06:29:00
#>  5 abbyyR/      2019-06-25 04:30:00 04:30 -     abbyyR       2019-06-25 04:30:00
#>  6 abc/         2022-05-19 07:20:00 07:20 -     abc          2022-05-19 07:20:00
#>  7 ABCanalysis/ 2017-03-13 13:31:00 13:31 -     ABCanalysis  2017-03-13 13:31:00
#>  8 abcdeFBA/    2022-03-07 06:20:00 06:20 -     abcdeFBA     2022-03-07 06:20:00
#>  9 ABCExtremes/ 2015-06-19 11:26:00 11:26 -     ABCExtremes  2015-06-19 11:26:00
#> 10 abclass/     2022-05-28 18:10:00 18:10 -     abclass      2022-05-28 18:10:00
#> # … with 37,813 more rows, and 4 more variables: archive <lgl>,
#> #   first_date <dttm>, n_versions <int>, last_modified <dttm>
```

The evolution of the number of packages on CRAN since its beginning.

``` r
cran_monthly_package_number
#> # A tibble: 297 × 2
#>    date       number_packages
#>    <date>               <dbl>
#>  1 1997-10-08               1
#>  2 1997-11-08               1
#>  3 1997-12-08               1
#>  4 1998-01-08               2
#>  5 1998-02-08               2
#>  6 1998-03-08               3
#>  7 1998-04-08               5
#>  8 1998-05-08               6
#>  9 1998-06-08               6
#> 10 1998-07-08               7
#> # … with 287 more rows
```

### Scraping tools

#### CRAN

Both `cran_packages_history` and `cran_monthly_package_number` datasets
are generated by the function `scrape_cran()`. The scraping process is
quite time consuming and relies on the `{furrr}` package to scrape the
CRAN pages asynchronously.

``` r
future::plan(future::multisession)
scrape_cran_history()
```

#### MRAN

`{cranology}` also includes the `get_package_number_mran()` function to
more quickly get the number of packages that were available on CRAN at
any given date.

``` r
dates <- seq(
  from = as.Date("2018-04-10", "%Y-%m-%d"), 
  by = "1 year", 
  length.out = 4
)
get_package_number_mran(dates)
#> Scraping MRAN...
#> Scraping number packages on: 2018-04-10
#> Scraping number packages on: 2019-04-10
#> Scraping number packages on: 2020-04-10
#> Scraping number packages on: 2021-04-10
#>         date number_packages
#> 1 2018-04-10           12418
#> 2 2019-04-10           14042
#> 3 2020-04-10           15507
#> 4 2021-04-10           17398
```

Be careful though as this will only work for dates posterior to
`2014-09-17` the day when MRAN was up online for the first time.

``` r
get_package_number_mran("2013-08-28")
#> Error: Some dates are anterior to MRAN launch:
#> 1: 2013-08-28
```

For earlier dates use `cran_monthly_package_number`. Here is a naïve
example:

``` r
date_before_mran <- as.Date("2013-08-28")

cran_monthly_package_number[
  min(
    which(
      cran_monthly_package_number$date >= date_before_mran
    )
  ), 
]
#> # A tibble: 1 × 2
#>   date       number_packages
#>   <date>               <dbl>
#> 1 2013-09-08            4906
```

## Acknowledgements

The `scrape_cran()` function is essentially a tidyversification of this
[github gist](https://gist.github.com/daroczig/3cf06d6db4be2bbe3368)
written by @daroczig.

## Code of Conduct

Please note that this project is released with a [Contributor Code of
Conduct](https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.


NEWS.md
# cranology 0.1.0

### Stabilise API

* `scrape_cran()`: function to generate both `cran_packages_history` and `cran_monthly_package_number` datasets.
* `get_package_number_mran()`: function to scrape number of available packages on MRAN.
* `update_montly_package_number()`: uses `get_package_number_mran()` to update datase `cran_monthly_package_number`.
* `plot_montly_package_number()`: function to plot `cran_monthly_package_number`.

### Misc

* Add Continous Integration
* Update README to a decent standard


R/scrape_mran.R

#' @importFrom lubridate ymd
#' @noRd
is_parsable_date <- function(date) {
  parsed_date <- ymd(date, quiet = TRUE)
  if ( is.na(parsed_date) ) {
    FALSE
  } else {
    TRUE
  }
}

date_error_message <- function() {
  paste0(
    "`dates` must be either:\n",
    "- a Date -> as.Date(Sys.time())\n",
    "- a character string of the form 'year-month-day': -> '2018-04-10'\n"
  )
}

#' @importFrom lubridate is.Date
#' @noRd
is_invalid_date <- function(date) {
  !is.Date(date) & is.character(date) & !is_parsable_date(date)
}

#' @noRd
is_invalid_dates <- function(dates) {
  vapply(
    X = dates,
    FUN = is_invalid_date,
    FUN.VALUE = logical(1),
    USE.NAMES = FALSE
  )
}

is_anterior_to_mran_launch <- function(valid_date) {
  as.Date("2014-09-17") > valid_date
}

#' @importFrom utils available.packages
#' @noRd
.get_package_number_mran <- function(date) {
  tryCatch(
    error = function(cnd) {
      message(
        sprintf(
          "Impossible to query MRAN for this particualar date: %s",
          date
        )
      )
      data.frame(
        date = date,
        n = NA
      )
    },
    {
      message(
        sprintf("Scraping number packages on: %s", date)
      )
      data.frame(
        date = date,
        number_packages = nrow(
          available.packages(
            repos = sprintf("https://mran.microsoft.com/snapshot/%s", date)
          )
        )
      )
    }
  )
}



#' Get number of package on CRAN on a given date using MRAN
#'
#' This function queries MRAN to retrieve the number of package on CRAN
#' on a given date.
#'
#' @param dates A vector of dates. Either a character vector of the form "yyyy-mm-dd"
#' or a vector of class "Date". All dates must be posterior to "2014-09-17", the
#' day of MRAN first CRAN snapshot.
#'
#' @param parallelize A logical. If TRUE {furrr} is used to asynchronously scrap
#' mran.
#'
#' @return A data.frame with two columns `date` and `n` the number of packages on CRAN
#' at that given `date`.
#'
#' @importFrom purrr map_dfr
#' @importFrom furrr future_map_dfr
#'
#' @export
#'
#' @name mran
#'
#' @examples
#' get_package_number_mran(c("2018-04-10", "2020-03-19"))

get_package_number_mran <- function(dates, parallelize = FALSE) {

  invalid_dates <- is_invalid_dates(dates)
  if ( any(invalid_dates) ) {
    stop(
      "Some dates are not valid dates:\n",
      paste(
        paste0(which(invalid_dates), ": ", dates[invalid_dates]),
        collapse = "\n"
      ),
      "\n",
      date_error_message(),
      call. = FALSE
    )
  }

  anterior_to_mran_launch <- is_anterior_to_mran_launch(dates)
  if ( any(anterior_to_mran_launch) ) {
    stop(
      "Some dates are anterior to MRAN launch:\n",
      paste(
        paste0(which(anterior_to_mran_launch), ": ", dates[anterior_to_mran_launch]),
        collapse = "\n"
      ),
      call. = FALSE
    )
  }

  message("Scraping MRAN...")

  if ( isTRUE(parallelize) ) {
    future_map_dfr(
      dates,
      .get_package_number_mran
    )
  } else {
    map_dfr(
      dates,
      .get_package_number_mran
    )
  }
}

#' Update `cran_monthly_package_number` dataset
#'
#' The creation `cran_monthly_package_number` using `scrape_cran()` is a long process
#' as theunderlying scrapping operations are time consuming. To more rapidly
#' update `cran_monthly_package_number` it is easier to rely on data from MRAN.
#' This what this function does. It uses `get_package_number_mran()` to quickly
#' update the dataset.
#'
#' @param cran_monthly_package_number_df A data.frame similar to
#' the `cran_monthly_package_number` dataset included within {cranology}.
#' @inheritParams mran
#'
#' @importFrom utils tail
#' @importFrom lubridate month days today
#'
#' @export
#' @examples
#' # Simulate `cran_monthly_package_number` update
#'date_lag <- 3
#'df <- cran_monthly_package_number[
#'  1:(nrow(cran_monthly_package_number) - date_lag),
#']
#'update_monthly_package_number(
#'  cran_monthly_package_number_df = df
#')
#'
update_monthly_package_number <- function(
    cran_monthly_package_number_df,
    parallelize = FALSE
) {
  first_date <- tail(
    cran_monthly_package_number_df,
    1
  )[["date"]] + month(1) - days(1)
  dates <- seq(
    from = first_date,
    to = today(),
    by = "1 month"
  )
  recent_months <- get_package_number_mran(
    dates = dates,
    parallelize = parallelize
  )
  rbind(
    cran_monthly_package_number_df,
    recent_months
  )
}
